smol/capp/generate/groundstate/solver.py

```diff
@@ -0,0 +1,553 @@
+        """
+        return self._canonicals.problem
+
+    @property
+    def variables(self):
+        """CVXPY variables for species on each active site.
+
+        Returns:
+            cvxpy.Variable.
+        """
+        return self._canonicals.variables
+
+    @property
+    def variable_indices(self):
+        """Indices of variables on each site if site is active.
+
+        Returns:
+            list of lists of int: each sub-list contains variable
+            indices on an active site.
+        """
+        return self._canonicals.variable_indices
+
+    @property
+    def objective_function(self):
+        """The objective function to be minimized.
+
+        Usually total energy per super-cell.
+
+        Returns:
+            cvxpy.Expression.
+        """
+        return self._canonicals.objective_function
+
+    @property
+    def constraints(self):
+        """Constraints to be satisfied in minimization.
+
+        Must include normalization per site. May include
+        charge balance and other constraints to composition.
+        In canonical ensemble, must include a constraint to
+        fix composition.
+
+        Returns:
+            List[cvxpy.Constraint]
+        """
+        return self._canonicals.constraints
+
+    @property
+    def auxiliary_variables(self):
+        """Auxiliary variables for linearizing the problem.
+
+        Returns:
+            cp.Variable.
+        """
+        return self._canonicals.auxiliary_variables
+
+    @property
+    def indices_in_auxiliary_products(self):
+        """Indices of variables in cluster terms corresponding to auxiliary variables.
+
+        Returns:
+            list of lists of int
+        """
+        return self._canonicals.indices_in_auxiliary_products
+
+    def solve(self):
+        """Solve the MIP problem.
+
+        Returns:
+            np.ndarray, float:
+                values of boolean variables in the ground-state solution, and
+                the minimized ground-state energy.
+        """
+        if (
+            self._ground_state_solution is not None
+            and self._ground_state_energy is not None
+        ):
+            warn("Ground state already solved before. Overwriting previous result!")
+        self.problem.solve(
+            solver=self.solver, warm_start=self.warm_start, **self.solver_options
+        )
+        if (
+            self.variables.value is not None
+            and self.objective_function.value is not None
+        ):
+            self._ground_state_solution = self.variables.value.astype(int)
+            self._ground_state_energy = self.objective_function.value
+        else:
+            warn(
+                "Ground state could not be solved! Try using another solver, or"
+                " fine tune solver options!"
+            )
+            self._ground_state_solution = None
+            self._ground_state_energy = None
+        return self._ground_state_solution, self._ground_state_energy
```

Comment:
I think `solve` should not return anything.

A user can access the solution in the attributes you have implemented. And my guess is that they will almost always only care about the ground state structure and energy

Reply:
OK. I can remove it. However, other @property attributes will be kept for the user's convenience.

Reply:
Removed explicit returns.